src/traces/mesh3d/defaults.js

```diff
@@ -34,18 +34,58 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
     }
 
     var coords = readComponents(['x', 'y', 'z']);
-    var indices = readComponents(['i', 'j', 'k']);
-
     if(!coords) {
         traceOut.visible = false;
         return;
     }
 
-    if(indices) {
-        // otherwise, convert all face indices to ints
-        indices.forEach(function(index) {
-            for(var i = 0; i < index.length; ++i) index[i] |= 0;
+    // three indices should be all provided or not
+    if(
+        (traceIn.i && (!traceIn.j || !traceIn.k)) ||
+        (traceIn.j && (!traceIn.k || !traceIn.i)) ||
+        (traceIn.k && (!traceIn.i || !traceIn.j))
+    ) {
+        traceOut.visible = false;
+        return;
+    }
+
+    // test for size of indices
+    if(
+        traceIn.i && Lib.isArrayOrTypedArray(traceIn.i) &&
+        traceIn.j && Lib.isArrayOrTypedArray(traceIn.j) &&
+        traceIn.k && Lib.isArrayOrTypedArray(traceIn.k)
+    ) {
+        if(traceIn.k.length !== 0 && (
+            traceIn.i.length !== traceIn.j.length ||
+            traceIn.j.length !== traceIn.k.length)) {
+            traceOut.visible = false;
+            return;
+        }
+    }
+
+    var allIndices = readComponents(['i', 'j', 'k']);
```

Comment:
This logic here belongs in `convert.js`. We should never loop over data arrays in `supplyDefaults`.

Reply:
Done in https://github.com/plotly/plotly.js/pull/3369/commits/a041e51366a6c5b281ec8f7e50026ce2acafbc94